ffi/src/lib.rs

```diff
@@ -143,15 +199,125 @@ fn batch(db: *mut Db, nkeys: usize, values: *const KeyValue) -> Result<Value, St
+    nkeys: usize,
+    values: *const KeyValue,
+) -> Value {
+    // Note: the id is guaranteed to be non-zero
+    // because we use an atomic counter that starts at 1.
+    propose_on_db(db, nkeys, values)
+        .map(|id| id.into())
+        .unwrap_or_else(|e| e.into())
+}
+
+/// cbindgen::ignore
+///
+/// This function is not exposed to the C API.
+/// Internal call for `fwd_propose_on_db` to remove error handling from the C API
+fn propose_on_db(
+    db: *const DatabaseHandle,
+    nkeys: usize,
+    values: *const KeyValue,
+) -> Result<ProposalId, String> {
+    let db = unsafe { db.as_ref() }.ok_or_else(|| String::from("db should be non-null"))?;
+
+    // Create a batch of operations to perform.
+    let mut batch = Vec::with_capacity(nkeys);
+    for i in 0..nkeys {
+        let kv = unsafe { values.add(i).as_ref() }
+            .ok_or_else(|| String::from("couldn't get key-value pair"))?;
+        if kv.value.len == 0 {
+            batch.push(DbBatchOp::DeleteRange {
+                prefix: kv.key.as_slice(),
+            });
+            continue;
+        }
+        batch.push(DbBatchOp::Put {
+            key: kv.key.as_slice(),
+            value: kv.value.as_slice(),
+        });
+    }
+    let proposal = db.propose_sync(batch).map_err(|e| e.to_string())?;
+    let proposal_id = next_id(); // Guaranteed to be non-zero
+    db.proposals.write().unwrap().insert(proposal_id, proposal);
+    Ok(proposal_id)
+}
+
+/// Commits a proposal to the database.
+///
+/// # Arguments
+///
+/// * `db` - The database handle returned by `open_db`
+/// * `proposal_id` - The ID of the proposal to commit
+///
+/// # Returns
+///
+/// A `Value` containing {0, null} if the commit was successful.
+/// A `Value` containing {0, "error message"} if the commit failed.
+///
+/// # Safety
+///
+/// This function is unsafe because it dereferences raw pointers.
+/// The caller must ensure that `db` is a valid pointer returned by `open_db`
+///
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn fwd_commit(db: *const DatabaseHandle, proposal_id: u32) -> Value {
+    commit(db, proposal_id)
+        .map(|e| e.into())
+        .unwrap_or_else(|e| e.into())
+}
+
+fn commit(db: *const DatabaseHandle, proposal_id: u32) -> Result<(), String> {
+    let db = unsafe { db.as_ref() }.ok_or_else(|| String::from("db should be non-null"))?;
+    let proposal = db
+        .proposals
+        .write()
+        .unwrap()
+        .remove(&proposal_id)
+        .ok_or_else(|| String::from("proposal not found"))?;
+    proposal.commit_sync().map_err(|e| e.to_string())
+}
+
 /// Get the root hash of the latest version of the database
 /// Don't forget to call `free_value` to free the memory associated with the returned `Value`.
 ///
+/// # Argument
+///
+/// * `db` - The database handle returned by `open_db`
+///
+/// # Returns
+///
+/// A `Value` containing the root hash of the database.
+/// A `Value` containing {0, "error message"} if the commit failed.
```

Comment:
again, not if the commit failed, something else is probably amiss...

Let's be clear about what happens when you call this on a database with nothing in it